3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,271 @@
+          this.creativeMessageRouters_[creativeId] =
+            this.creativeMessageRouters_[creativeId] ||
+            new AmpAnalytics3pCreativeMessageRouter(this, creativeId,
+              extraData);
+        });
+      });
+    this.iframeMessagingClient_.registerCallback(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.EVENT,
+      messageContainer => {
+        Object.entries(messageContainer.data).forEach(entry => {
+          const creativeId = entry[0];
+          const messages = entry[1];
+          dev().assert(messages.length > 0, 'Received empty events list');
+          dev().assert(this.creativeMessageRouters_[creativeId],
+            'Received event message prior to new creative message.' +
+            ' Discarding.');
+          this.creativeMessageRouters_[creativeId]
+            .receiveEventMessages(messages);
+        });
+        this.flushQueues_();
+      });
+    this.iframeMessagingClient_.sendMessage(
+      AMP_ANALYTICS_3P_MESSAGE_TYPE.READY);
+  }
+
+  /**
+   * Gets a handle to the window (or testing object that mocks window)
+   * @returns {!Window}
+   */
+  getWindow() {
+    return this.win_;
+  }
+
+  /**
+   * Getter to expose iframeMessagingClient to instances of
+   * AmpAnalytics3pCreativeMessageRouter
+   * @returns {!IframeMessagingClient}
+   */
+  getMessagingClient() {
+    return this.iframeMessagingClient_;
+  }
+
+  /**
+   * Instructs each AmpAnalytics3pCreativeMessageRouter instances to send its
+   * queued messages to its listener (if it has a listener yet).
+   * @private
+   */
+  flushQueues_() {
+    Object.entries(this.creativeMessageRouters_).forEach(
+      entry => {
+        entry[1].sendQueuedMessagesToListener();
+      });
+  }
+}
+
+new AmpAnalytics3pMessageRouter(window);
+
+/**
+ * Receives messages bound for this cross-domain iframe, from a particular
+ * creative. Also sends repsonse messages from the iframe meant for this
+ * particular creative.
+ */
+class AmpAnalytics3pCreativeMessageRouter {
+  /**
+   * @param {!AmpAnalytics3pMessageRouter} parent
+   * @param {!string} creativeId
+   * @param {string=} opt_extraData
+   */
+  constructor(parent, creativeId, opt_extraData) {
+    /** @private {!AmpAnalytics3pMessageRouter} */
+    this.parent_ = parent;
+
+    /** @private {!IframeMessagingClient} */
+    this.iframeMessagingClient_ = parent.getMessagingClient();
+
+    /**
+     * @private {!string}
+     */
+    this.creativeId_ = creativeId;
+
+    /**
+     * @private {?string}
+     */
+    this.extraData_ = opt_extraData;
+
+    /**
+     * private {!Array<!AmpAnalytics3pEvent>}
+     */
+    this.receivedMessagesQueue_ = [];
+
+    /**
+     * private {?function(!Array<!AmpAnalytics3pEvent>)=}
+     */
+    this.eventListener_ = null;
+
+    if (this.parent_.getWindow().onNewAmpAnalyticsInstance) {
+      this.parent_.getWindow().onNewAmpAnalyticsInstance(this);
+    } else {
+      dev().error(TAG_, 'Vendor page must implement onNewAmpAnalyticsInstance' +
```

Comment:
Ideally you could indicate which vendor this is as there could be multiple on the page.  Also the message is not really true... the vendor must have declared new amp analytics instance prior to receiving new creative message but I suppose that's getting into implementation details :)